src/main/java/ch/njol/skript/effects/EffVehicle.java

```diff
@@ -15,93 +10,86 @@
-	@Nullable
-	private Expression<?> vehicles;
-	
-	@SuppressWarnings({"unchecked", "null"})
+
+	private @Nullable Expression<Entity> passengers;
+	private @Nullable Expression<?> vehicles;
+
 	@Override
-	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		//noinspection unchecked
 		passengers = matchedPattern == 2 ? null : (Expression<Entity>) exprs[0];
 		vehicles = matchedPattern == 1 ? null : exprs[exprs.length - 1];
-		if (!PassengerUtils.hasMultiplePassenger() && passengers != null && vehicles != null && !passengers.isSingle() && vehicles.isSingle() && Entity.class.isAssignableFrom(vehicles.getReturnType()))
-			Skript.warning("An entity can only have one passenger");
 		return true;
 	}
 	
 	@Override
-	protected void execute(final Event e) {
-		final Expression<?> vehicles = this.vehicles;
-		final Expression<Entity> passengers = this.passengers;
+	protected void execute(Event event) {
+		// matchedPattern = 1
 		if (vehicles == null) {
 			assert passengers != null;
-			for (final Entity p : passengers.getArray(e))
-				p.leaveVehicle();
+			for (Entity passenger : passengers.getArray(event))
+				passenger.leaveVehicle();
 			return;
 		}
+		// matchedPattern = 2
 		if (passengers == null) {
-			assert vehicles != null;
-			for (final Object v : vehicles.getArray(e))
-				((Entity) v).eject();
+			for (Object vehicle : vehicles.getArray(event))
+				((Entity) vehicle).eject();
 			return;
 		}
-		final Object[] vs = vehicles.getArray(e);
-		if (vs.length == 0)
-			return;
-		final Entity[] ps = passengers.getArray(e);
-		if (ps.length == 0)
+		// matchedPattern = 0
+		Entity[] passengersArray = passengers.getArray(event);
+		if (passengersArray.length == 0)
 			return;
-		for (final Object v : vs) {
-			if (v instanceof Entity) {
-				((Entity) v).eject();
-				for (Entity p : ps){
-					assert p != null;
-					p.leaveVehicle();
-					PassengerUtils.addPassenger((Entity)v, p); //For 1.9 and lower, it will only set the last one.
-				}
-			} else {
-				for (final Entity p : ps) {
-					assert p != null : passengers;
-					final Entity en = ((EntityData<?>) v).spawn(p.getLocation());
-					if (en == null)
-						return;
-					PassengerUtils.addPassenger(en, p);
-				}
+		Object vehicleObject = vehicles.getSingle(event);
+		if (vehicleObject instanceof Entity vehicleEntity) {
+			for (Entity passenger : passengersArray) {
+				assert passenger != null;
+				if (passenger == vehicleEntity)
+					continue;
+				passenger.leaveVehicle();
+				vehicleEntity.addPassenger(passenger);
+			}
+		} else if (vehicleObject instanceof EntityData<?> vehicleData) {
+			for (Entity passenger : passengersArray) {
+				assert passenger != null;
+				Entity vehicleEntity = vehicleData.spawn(passenger.getLocation());
+				if (vehicleEntity == null)
+					return;
```

Comment:
shouldn't we make the passenger leave their current vehicle here too or is that unnecessary?